### PR TITLE
[BlockStore] Dequeue all items from the async queue on dispose

### PR DESCRIPTION
**Pull request by fassadlr**

### CODE REVIEW

File: src/Stratis.Bitcoin/Utilities/AsyncQueue.cs [hunk 1]
```diff
@@ -47,6 +47,18 @@ private class AsyncContext
         /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
         private readonly Queue<T> items;
 
+        /// <summary>The amount of items in the queue.</summary>
+        public int ItemCount
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Utilities/AsyncQueue.cs]: Why not just `Count`?

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs [hunk 2]
```diff
@@ -394,6 +394,20 @@ private async Task DequeueBlocksContinuouslyAsync()
...
+
+        /// <summary>
+        /// Ensures that any blocks queued in <see cref="blocksQueue"/> gets added to <see cref="batch"/>
+        /// so that it can be persisted on dispose.
+        /// </summary>
+        private async Task SaveBlocksInBatchAndAsyncQueueAsync()
+        {
+            while (this.blocksQueue.ItemCount > 0)
+            {
+                this.batch.Add(await this.blocksQueue.DequeueAsync());
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs]: configureawait(false)

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs [hunk 3]
```diff
@@ -394,6 +394,20 @@ private async Task DequeueBlocksContinuouslyAsync()
                 }
             }
 
+            await SaveBlocksInBatchAndAsyncQueueAsync();
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs]: `this.`

**Comment by noescape00** [comment]
I think it will also be a good idea to add a check to `AddToPending` that cancellation token is not cancelled yet to avoid items consumption during disposure. Throw if we are attempting to add a new item after BSQ's token is cancelled.

**Comment by noescape00** [comment]
I'm not sure if count should be added to the async queue. I remember we discussed that with @Aprogiena but I don't remember his point on why it was a bad idea.

File: src/Stratis.Bitcoin/Utilities/AsyncQueue.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Utilities/AsyncQueue.cs]: Fixed 👍

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs]: Fixed.

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs]: Fixed

**Comment by noescape00** [comment]
>@noescape00 will it make more sense to make TryDequeue a public method?

That can work as well as long as it's thread safe, yeah.

**Comment by dangershony** [comment]
Yeah it is thread safe internally already

**Pull request merged**